File: Components/WinForms/CS/ChartsMainDemo/Modules/ChartElements/Legend.cs
using System;
using DevExpress.Utils.Menu;
using DevExpress.XtraCharts.Demos.PointLineViews;
using DevExpress.XtraEditors;

namespace DevExpress.XtraCharts.Demos.ChartElements {
	public partial class LegendDemo : ChartDemoBase2D {
		ChartControl chart;
        DXSubMenuItem menuLegendVertItem;
        DXSubMenuItem menuLegendHorItem;

		public override ChartControl ChartControl { get { return chart; } }

		public LegendDemo() {
			InitializeComponent();
		}

		void AddDirectionClickHandlers(DXMenuItem directionItem) {
            DXSubMenuItem item = (DXSubMenuItem)directionItem;
            item.Items[0].Click += LegendDirectionItem0Click;
            item.Items[1].Click += LegendDirectionItem1Click;
            item.Items[2].Click += LegendDirectionItem2Click;
            item.Items[3].Click += LegendDirectionItem3Click;
		}
        void LegendDirectionItem0Click(object sender, EventArgs e) {
            cbEquallySpacedItems.Enabled = false;
            cbDirection.SelectedIndex = 0;
        }
        void LegendDirectionItem1Click(object sender, EventArgs e) {
            cbEquallySpacedItems.Enabled = false;
            cbDirection.SelectedIndex = 1;
        }
        void LegendDirectionItem2Click(object sender, EventArgs e) {
            cbEquallySpacedItems.Enabled = true;
            cbDirection.SelectedIndex = 2;
        }
        void LegendDirectionItem3Click(object sender, EventArgs e) {
            cbEquallySpacedItems.Enabled = true;
            cbDirection.SelectedIndex = 3;
        }
        void AddVertAlignmentClickHandlers(DXMenuItem vertItem) {
            menuLegendVertItem = (DXSubMenuItem)vertItem;
            for (int i = 0; i < menuLegendVertItem.Items.Count; i++)
                menuLegendVertItem.Items[i].Click += LegendVertAlignmentClick;
		}
		void AddHorAlignmentClickHandlers(DXMenuItem horItem) {
            menuLegendHorItem = (DXSubMenuItem)horItem;
            for (int i = 0; i < menuLegendHorItem.Items.Count; i++)
                menuLegendHorItem.Items[i].Click += LegendHorAlignmentClick;
		}
        void LegendVertAlignmentClick(object sender, EventArgs e) {
            cbVertAlign.SelectedIndex = GetCheckedAlign((DXMenuItem)sender, menuLegendVertItem);
        }
        void LegendHorAlignmentClick(object sender, EventArgs e) {
            cbHorAlign.SelectedIndex = GetCheckedAlign((DXMenuItem)sender, menuLegendHorItem);
        }
        string GetValueWithoutPercent(string val) {
			return val.Trim(new char[] { '%' });
		}
		string AddPercentToString(string val) {
			return val + "%";
		}
		void UpdateEquallySpacedItemsEnable() {
			cbEquallySpacedItems.Enabled = ChartControl.Legend.Direction == LegendDirection.LeftToRight ||
				ChartControl.Legend.Direction == LegendDirection.RightToLeft;
		}
		int GetCheckedAlign(DXMenuItem sender, DXSubMenuItem item) {
            if (item != null) {
                for (int index = 0; index < item.Items.Count; index++)
                    if (object.ReferenceEquals(item.Items[index], sender))
                        return index;
            }
			return -1;
		}
		void InitSeries() {
			Random random = new Random();
			for (int i = 0; i < chart.Series.Count; i++) {
				chart.Series[i].Points.Clear();
				chart.Series[i].Points.AddRange(RandomPointCalculator.Calculate(random, 5, 1, 10, 1, 10));
			}
		}
		void chart_VisibleChanged(object sender, EventArgs e) {
			InitSeries();
		}
		void itemVertAlignment_Click(object sender, EventArgs e) {
			cbVertAlign.SelectedIndex = (int)ChartControl.Legend.AlignmentVertical;
		}
        void UpdateLegendMargins() {
            LegendAlignmentHorizontal hAlignment = chart.Legend.AlignmentHorizontal;
            LegendAlignmentVertical vAlignment = chart.Legend.AlignmentVertical;
            if ((object.Equals(hAlignment, LegendAlignmentHorizontal.Left) || object.Equals(hAlignment, LegendAlignmentHorizontal.Center) || object.Equals(hAlignment, LegendAlignmentHorizontal.Right))
                && (object.Equals(vAlignment, LegendAlignmentVertical.Bottom) || object.Equals(vAlignment, LegendAlignmentVertical.Center) || object.Equals(vAlignment, LegendAlignmentVertical.Top)))
                chart.Legend.Margins.All = 10;
            else
                chart.Legend.Margins.All = 5;
        }
		void itemHorAlignment_Click(object sender, EventArgs e) {
			cbHorAlign.SelectedIndex = (int)ChartControl.Legend.AlignmentHorizontal;
		}
		void cbEquallySpacedItems_CheckedChanged(object sender, EventArgs e) {
			if (chart.Legend.Direction == LegendDirection.LeftToRight || 
				chart.Legend.Direction == LegendDirection.RightToLeft)
					ChartControl.Legend.EquallySpacedItems = cbEquallySpacedItems.Checked;
		}
		void cmbMaxHorPercentage_SelectedIndexChanged(object sender, EventArgs e) {
			string value = GetValueWithoutPercent(cmbMaxHorPercentage.Text);
			ChartControl.Legend.MaxHorizontalPercentage = Convert.ToDouble(value);
		}
		void cmbMaxVertPercentage_SelectedIndexChanged(object sender, EventArgs e) {
			string value = GetValueWithoutPercent(cmbMaxVertPercentage.Text);
			ChartControl.Legend.MaxVerticalPercentage = Convert.ToDouble(value);
		}
		void cbDirection_SelectedIndexChanged(object sender, EventArgs e) {
			ComboBoxEdit edit = sender as ComboBoxEdit;
			if (edit != null) {
    			ChartControl.Legend.Direction = (LegendDirection)edit.SelectedIndex;
	    		if (edit.SelectedIndex == 0 || edit.SelectedIndex == 1)
		    		cbEquallySpacedItems.Enabled = false;
			    else
				    cbEquallySpacedItems.Enabled = true;
            }
		}
		void cbHorAlign_SelectedIndexChanged(object sender, EventArgs e) {
			ComboBoxEdit edit = sender as ComboBoxEdit;
            if (edit != null) {
                ChartControl.Legend.AlignmentHorizontal = (LegendAlignmentHorizontal)edit.SelectedIndex;
                UpdateLegendMargins();
            }
		}
		void cbVertAlign_SelectedIndexChanged(object sender, EventArgs e) {
			ComboBoxEdit edit = (sender) as ComboBoxEdit;
            if (edit != null) {
                ChartControl.Legend.AlignmentVertical = (LegendAlignmentVertical)edit.SelectedIndex;
                UpdateLegendMargins();
            }
		}
		
        protected override DXPopupMenu ConstructPopupMenu(object obj, ChartControl chartControl) {
			DXPopupMenu menu = base.ConstructPopupMenu(obj, chartControl);
			if (obj is Legend) {
				AddHorAlignmentClickHandlers(menu.Items[4]);
				AddVertAlignmentClickHandlers(menu.Items[5]);
				AddDirectionClickHandlers(menu.Items[6]);
			}
			return menu;
		}
		protected override void InitControls() {
			base.InitControls();
            CheckEditShowLabelsVisible = false;
			UpdateEquallySpacedItemsEnable();
			cbEquallySpacedItems.Checked = chart.Legend.EquallySpacedItems;
			cmbMaxHorPercentage.Text = AddPercentToString(ChartControl.Legend.MaxHorizontalPercentage.ToString());
			cmbMaxVertPercentage.Text = AddPercentToString(ChartControl.Legend.MaxVerticalPercentage.ToString());
			cbDirection.SelectedIndex = (int)ChartControl.Legend.Direction;
			cbHorAlign.SelectedIndex = (int)ChartControl.Legend.AlignmentHorizontal;
			cbVertAlign.SelectedIndex = (int)ChartControl.Legend.AlignmentVertical;
		}
        
        public override void UpdateControls() {
            base.UpdateControls();
            cbVertAlign.SelectedIndex = (int)chart.Legend.AlignmentVertical;
            cbHorAlign.SelectedIndex = (int)chart.Legend.AlignmentHorizontal;
            cbDirection.SelectedIndex = (int)chart.Legend.Direction;
            cbEquallySpacedItems.Checked = chart.Legend.EquallySpacedItems;
        }
	}
}

File: Components/WinForms/VB/ChartsMainDemo/Modules/ChartElements/Legend.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraCharts.Demos.PointLineViews
Imports DevExpress.XtraEditors

Namespace DevExpress.XtraCharts.Demos.ChartElements
	Partial Public Class LegendDemo
		Inherits ChartDemoBase2D
		Private WithEvents chart As ChartControl
		Private menuLegendVertItem As DXSubMenuItem
		Private menuLegendHorItem As DXSubMenuItem

		Public Overrides ReadOnly Property ChartControl() As ChartControl
			Get
				Return chart
			End Get
		End Property

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub AddDirectionClickHandlers(ByVal directionItem As DXMenuItem)
			Dim item As DXSubMenuItem = CType(directionItem, DXSubMenuItem)
			AddHandler item.Items(0).Click, AddressOf LegendDirectionItem0Click
			AddHandler item.Items(1).Click, AddressOf LegendDirectionItem1Click
			AddHandler item.Items(2).Click, AddressOf LegendDirectionItem2Click
			AddHandler item.Items(3).Click, AddressOf LegendDirectionItem3Click
		End Sub
		Private Sub LegendDirectionItem0Click(ByVal sender As Object, ByVal e As EventArgs)
			cbEquallySpacedItems.Enabled = False
			cbDirection.SelectedIndex = 0
		End Sub
		Private Sub LegendDirectionItem1Click(ByVal sender As Object, ByVal e As EventArgs)
			cbEquallySpacedItems.Enabled = False
			cbDirection.SelectedIndex = 1
		End Sub
		Private Sub LegendDirectionItem2Click(ByVal sender As Object, ByVal e As EventArgs)
			cbEquallySpacedItems.Enabled = True
			cbDirection.SelectedIndex = 2
		End Sub
		Private Sub LegendDirectionItem3Click(ByVal sender As Object, ByVal e As EventArgs)
			cbEquallySpacedItems.Enabled = True
			cbDirection.SelectedIndex = 3
		End Sub
		Private Sub AddVertAlignmentClickHandlers(ByVal vertItem As DXMenuItem)
			menuLegendVertItem = CType(vertItem, DXSubMenuItem)
			For i As Integer = 0 To menuLegendVertItem.Items.Count - 1
				AddHandler menuLegendVertItem.Items(i).Click, AddressOf LegendVertAlignmentClick
			Next i
		End Sub
		Private Sub AddHorAlignmentClickHandlers(ByVal horItem As DXMenuItem)
			menuLegendHorItem = CType(horItem, DXSubMenuItem)
			For i As Integer = 0 To menuLegendHorItem.Items.Count - 1
				AddHandler menuLegendHorItem.Items(i).Click, AddressOf LegendHorAlignmentClick
			Next i
		End Sub
		Private Sub LegendVertAlignmentClick(ByVal sender As Object, ByVal e As EventArgs)
			cbVertAlign.SelectedIndex = GetCheckedAlign(CType(sender, DXMenuItem), menuLegendVertItem)
		End Sub
		Private Sub LegendHorAlignmentClick(ByVal sender As Object, ByVal e As EventArgs)
			cbHorAlign.SelectedIndex = GetCheckedAlign(CType(sender, DXMenuItem), menuLegendHorItem)
		End Sub
		Private Function GetValueWithoutPercent(ByVal val As String) As String
			Return val.Trim(New Char() { "%"c })
		End Function
		Private Function AddPercentToString(ByVal val As String) As String
			Return val & "%"
		End Function
		Private Sub UpdateEquallySpacedItemsEnable()
			cbEquallySpacedItems.Enabled = ChartControl.Legend.Direction = LegendDirection.LeftToRight OrElse ChartControl.Legend.Direction = LegendDirection.RightToLeft
		End Sub
		Private Function GetCheckedAlign(ByVal sender As DXMenuItem, ByVal item As DXSubMenuItem) As Integer
			If item IsNot Nothing Then
				For index As Integer = 0 To item.Items.Count - 1
					If Object.ReferenceEquals(item.Items(index), sender) Then
						Return index
					End If
				Next index
			End If
			Return -1
		End Function
		Private Sub InitSeries()
			Dim random As New Random()
			For i As Integer = 0 To chart.Series.Count - 1
				chart.Series(i).Points.Clear()
				chart.Series(i).Points.AddRange(RandomPointCalculator.Calculate(random, 5, 1, 10, 1, 10))
			Next i
		End Sub
		Private Sub chart_VisibleChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chart.VisibleChanged
			InitSeries()
		End Sub
		Private Sub itemVertAlignment_Click(ByVal sender As Object, ByVal e As EventArgs)
			cbVertAlign.SelectedIndex = CInt(Fix(ChartControl.Legend.AlignmentVertical))
		End Sub
		Private Sub UpdateLegendMargins()
			Dim hAlignment As LegendAlignmentHorizontal = chart.Legend.AlignmentHorizontal
			Dim vAlignment As LegendAlignmentVertical = chart.Legend.AlignmentVertical
			If (Object.Equals(hAlignment, LegendAlignmentHorizontal.Left) OrElse Object.Equals(hAlignment, LegendAlignmentHorizontal.Center) OrElse Object.Equals(hAlignment, LegendAlignmentHorizontal.Right)) AndAlso (Object.Equals(vAlignment, LegendAlignmentVertical.Bottom) OrElse Object.Equals(vAlignment, LegendAlignmentVertical.Center) OrElse Object.Equals(vAlignment, LegendAlignmentVertical.Top)) Then
				chart.Legend.Margins.All = 10
			Else
				chart.Legend.Margins.All = 5
			End If
		End Sub
		Private Sub itemHorAlignment_Click(ByVal sender As Object, ByVal e As EventArgs)
			cbHorAlign.SelectedIndex = CInt(Fix(ChartControl.Legend.AlignmentHorizontal))
		End Sub
		Private Sub cbEquallySpacedItems_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbEquallySpacedItems.CheckedChanged
			If chart.Legend.Direction = LegendDirection.LeftToRight OrElse chart.Legend.Direction = LegendDirection.RightToLeft Then
					ChartControl.Legend.EquallySpacedItems = cbEquallySpacedItems.Checked
			End If
		End Sub
		Private Sub cmbMaxHorPercentage_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cmbMaxHorPercentage.SelectedIndexChanged
			Dim value As String = GetValueWithoutPercent(cmbMaxHorPercentage.Text)
			ChartControl.Legend.MaxHorizontalPercentage = Convert.ToDouble(value)
		End Sub
		Private Sub cmbMaxVertPercentage_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cmbMaxVertPercentage.SelectedIndexChanged
			Dim value As String = GetValueWithoutPercent(cmbMaxVertPercentage.Text)
			ChartControl.Legend.MaxVerticalPercentage = Convert.ToDouble(value)
		End Sub
		Private Sub cbDirection_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbDirection.SelectedIndexChanged
			Dim edit As ComboBoxEdit = TryCast(sender, ComboBoxEdit)
			If edit IsNot Nothing Then
				ChartControl.Legend.Direction = CType(edit.SelectedIndex, LegendDirection)
				If edit.SelectedIndex = 0 OrElse edit.SelectedIndex = 1 Then
					cbEquallySpacedItems.Enabled = False
				Else
					cbEquallySpacedItems.Enabled = True
				End If
			End If
		End Sub
		Private Sub cbHorAlign_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbHorAlign.SelectedIndexChanged
			Dim edit As ComboBoxEdit = TryCast(sender, ComboBoxEdit)
			If edit IsNot Nothing Then
				ChartControl.Legend.AlignmentHorizontal = CType(edit.SelectedIndex, LegendAlignmentHorizontal)
				UpdateLegendMargins()
			End If
		End Sub
		Private Sub cbVertAlign_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbVertAlign.SelectedIndexChanged
			Dim edit As ComboBoxEdit = TryCast((sender), ComboBoxEdit)
			If edit IsNot Nothing Then
				ChartControl.Legend.AlignmentVertical = CType(edit.SelectedIndex, LegendAlignmentVertical)
				UpdateLegendMargins()
			End If
		End Sub

		Protected Overrides Function ConstructPopupMenu(ByVal obj As Object, ByVal chartControl As ChartControl) As DXPopupMenu
			Dim menu As DXPopupMenu = MyBase.ConstructPopupMenu(obj, chartControl)
			If TypeOf obj Is Legend Then
				AddHorAlignmentClickHandlers(menu.Items(4))
				AddVertAlignmentClickHandlers(menu.Items(5))
				AddDirectionClickHandlers(menu.Items(6))
			End If
			Return menu
		End Function
		Protected Overrides Sub InitControls()
			MyBase.InitControls()
			CheckEditShowLabelsVisible = False
			UpdateEquallySpacedItemsEnable()
			cbEquallySpacedItems.Checked = chart.Legend.EquallySpacedItems
			cmbMaxHorPercentage.Text = AddPercentToString(ChartControl.Legend.MaxHorizontalPercentage.ToString())
			cmbMaxVertPercentage.Text = AddPercentToString(ChartControl.Legend.MaxVerticalPercentage.ToString())
			cbDirection.SelectedIndex = CInt(Fix(ChartControl.Legend.Direction))
			cbHorAlign.SelectedIndex = CInt(Fix(ChartControl.Legend.AlignmentHorizontal))
			cbVertAlign.SelectedIndex = CInt(Fix(ChartControl.Legend.AlignmentVertical))
		End Sub

		Public Overrides Sub UpdateControls()
			MyBase.UpdateControls()
			cbVertAlign.SelectedIndex = CInt(Fix(chart.Legend.AlignmentVertical))
			cbHorAlign.SelectedIndex = CInt(Fix(chart.Legend.AlignmentHorizontal))
			cbDirection.SelectedIndex = CInt(Fix(chart.Legend.Direction))
			cbEquallySpacedItems.Checked = chart.Legend.EquallySpacedItems
		End Sub
	End Class
End Namespace
